engine/cache/service.go

```diff
@@ -109,28 +121,56 @@ func (r UpdateCacheRecordsRequest) String() string {
 	return string(b)
 }
 
-//nolint:revive
+// CacheKeyID is the ID of a cache record.
+//
+// We use these to connect records together, instead of using the array-indexed
+// records format that buildkit uses.
+type CacheKeyID string
+
+// CacheKey is the part of a CacheRecord that contains the results.
+//
+// TODO: rename this at some point - a cache key should just be the
+// ID/digest/etc, and shouldn't have associated results.
 type CacheKey struct {
-	ID      string
+	// ID is the unique identifier for a record.
+	ID CacheKeyID
+	// Results are all the results that are connected to this record.
 	Results []Result
 }
 
+// Link is connects two CacheKeys together.
+//
+// It's similar to upstream's CacheInfoLink. And CacheChain's link struct.
```

Comment:
A fun note, something that I think is actually an issue.

We don't have an `Output` digest here. I think the fact that it works without it is more coincidence than anything else (since I *think* ID is "unique-enough"). That said, if we refactor the internals of magicache to depend less on the stability of IDs (which we need to do to resolve some weird caching bugs), then I think we might need to grab `Output`.

:cry:

Reply:
The `Output` is inherently digested into the digest set in `Link.Digest`, the idea was that was all we needed (but obviously it's possible we may actually need it to fix the various issues).